ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+// Unsubscribe removes previously registered channel for the given message names.
+func (c *RealtimeChannel) Unsubscribe(ch chan *proto.Message, names ...string) {
+	if ch == nil {
+		panic("ably: RealtimeChannel.Unsubscribe using nil channel")
+	}
+	if len(names) == 0 {
+		names = []string{""}
+	}
 	c.listenMtx.Lock()
-	delete(c.listeners[event], ch)
-	if len(c.listeners[event]) == 0 {
-		delete(c.listeners, event)
+	for _, name := range names {
+		delete(c.listen[name], ch)
+		if len(c.listen[name]) == 0 {
+			delete(c.listen, name)
+		}
 	}
 	c.listenMtx.Unlock()
-	close(ch)
 }
 
-func (c *RealtimeChannel) Publish(name string, data string) error {
-	c.Attach()
+// On relays request channel states to c; on state transition
+// connection will not block sending to c - the caller must ensure the incoming
+// values are read at proper pace or the c is sufficiently buffered.
+//
+// If no states are given, c is registered for all of them.
+// If c is nil, the method panics.
+// If c is alreadt registered, its state set is expanded.
+func (c *RealtimeChannel) On(ch chan<- State, states ...int) {
+	c.state.on(ch, states...)
+}
+
+// Off removes c from listetning on the given channel state transitions.
+//
+// If no states are given, c is removed for all of the connection's states.
+// If c is nil, the method panics.
+func (c *RealtimeChannel) Off(ch chan<- State, states ...int) {
+	c.state.off(ch, states...)
+}
+
+// Publish publishes a message on the channel, which is send on separate
+// goroutine. Publish does not block.
+//
+// If listen channel is non-nil it will be used to notify whether the message
+// was published succesfully with nil error value; when the operation failed,
+// the error that caused the failure is going to be sent on the listen channel.
+//
+// This implicitly attaches the channel if it's not already attached.
+func (c *RealtimeChannel) Publish(name string, data string, listen chan<- error) error {
+	return c.PublishAll([]*proto.Message{{Name: name, Data: data}}, listen)
+}
+
+// PublishAll publishes all given messages on the channel at once.
+// PublishAll does not block.
+//
+// If listen channel is non-nil it will be used to notify whether the messages
+// were published succesfully with nil error value; when the operation failed,
+// the error that caused the failure will be sent on the listen channel.
+//
+// This implicitly attaches the channel if it's not already attached.
+func (c *RealtimeChannel) PublishAll(messages []*proto.Message, listen chan<- error) error {
+	if err := c.Attach(); err != nil {
+		return err
+	}
 	msg := &proto.ProtocolMessage{
-		Action:  proto.ActionMessage,
-		Channel: c.Name,
-		Messages: []*proto.Message{
-			{Name: name, Data: data},
-		},
+		Action:   proto.ActionMessage,
+		Channel:  c.Name,
+		Messages: messages,
 	}
-	return c.client.send(msg)
+	switch c.State() {
+	case StateChanInitialized, StateChanAttaching:
+		if c.client.opts.NoQueueing {
```

Comment:
Fixed.

I misunderstood how ClientOptions.Queueing is meant to work - this param is used to enable/disable queueing on the connection object, disabling it has no influence on channel, which is always queueing.